liferay-frontend/liferay-portal
# LPS-129801 Move useLiferayState into frontend-js-react-web

Nominal goal here is to "put things where they belong" (ie. move the `useLiferayState` React hook into `frontend-js-react-web`). The "real" objective though is to show that we can support cross-project TypeScript in our build and get the benefits of global type-checking, editor/IDE integration, and so on, without unduly impacting build performance. So, the major changes are:

- Port `frontend-js-react-web` to TypeScript.
- Update to [`@liferay/npm-scripts` v42.0.0](https://github.com/liferay/liferay-frontend-projects/releases/tag/npm-scripts%2Fv42.0.0) for build support.

Lots of details explained in:

- https://github.com/liferay/liferay-frontend-projects/pull/478
- https://github.com/liferay-frontend/liferay-portal/pull/942

But the high-level summary of how the TS build works can be found [in our guidelines](https://github.com/liferay/liferay-frontend-projects/blob/f5917d11d60d5b17ac3845e247e48ac87136ed22/projects/npm-tools/packages/npm-scripts/src/typescript/README.md).

In short, the TypeScript compiler (`tsc`) can support our modular project layout by using a feature called "project references". This means each project needs only worry about its own types (stored in the "types/" folder), so gets the safety of global type-checking across liferay-portal, but also the fast builds that come with building each project in isolation. In order to get the full benefit of editor integration with the type-system (eg. autocomplete, inline documentation, "go-to-definition" etc), we must manage and commit the "tsconfig.json" files in the root of each project. All of this is abstracted away inside `@liferay/npm-scripts` and up-to-date-ness is enforced during `ci:test:sf`.

Test plan:

1. Run `yarn run liferay-npm-scripts types` and see it prime all the config files.
2. Run `yarn run build` in TS-using modules and see them build. They emit a warning when out of date.
3. When clean, see that `yarn run build` emits no warning because everything is up-to-date.
4. See `ci:test:sf` correctly fail/pass on the test PR, linked above.
5. Manual testing of running `liferay-portal` instance showing that [this Clay sample using the new state management API](https://gist.github.com/wincent/2d1e822e6a3478488a5036fcaff1f30b) (still) works.

---

This is an evolution of [#925](https://github.com/liferay-frontend/liferay-portal/pull/925), sending this as a separate PR because we have a new LPS now, and I wanted to rename the branch too for consistency.

Motivation here is to "put things where they belong" but also use this as a vehicle to prove that cross-project TypeScript compilation can work.

- [x] Show that cross-module build completes without errors.
- [x] Show that cross-module built product works at runtime.
- [x] Show that cross-module tests involving TS work (needed [this `npm-scripts` PR](https://github.com/liferay/liferay-frontend-projects/pull/468)).
- [x] Port ~all~ most of frontend-js-react-web to TS (to avoid ugly `@ts-ignore` directives).
- [x] Port rest of frontend-js-react-web to TS (ie. tests, and "render.tsx", but that is going to require more changes to `npm-scripts`).
- [x] Move `.d.ts` files into final (committed) locations (although note in the corresponding commit messages that we have options here and I'm just showing one or two of them).
- [x] Teach `npm-scripts` to ensure type dependencies are built first and/or not stale.
- [x] Delegate management of `tsconfig.json` files to `npm-scripts`.
- [x] Roll up TS vs ESLint tweaks (`no-use-before-define`) into `npm-scripts` or similar (see: https://github.com/liferay/liferay-frontend-projects/pull/471).
- [x] Roll up Jest tweaks into `npm-scripts` or similar (see: https://github.com/liferay/liferay-frontend-projects/pull/473).
- [x] Pull in `@types/react-dom` via `npm-scripts`, probably... (see: https://github.com/liferay/liferay-frontend-projects/pull/472).
- [ ] Declare global `Liferay` types centrally. [Will deal with this separately as it might be fiddly]

~This is probably going to sit here over the weekend because me next steps are going to take me to `npm-scripts` — I'll keep this as a draft in the meantime.~

~P.S. Sorry for the huge commit list — I had to base on top of Brian's remote in order to get latest `npm-scripts` version.~
